[Step5] 자동차 경주(리팩터링)

kyuwon53 opened the pull request:
안녕하세요 ! 🚀 

꼼꼼한 [피드백](https://github.com/next-step/kotlin-racingcar/pull/1655) 감사드립니다 💯 덕분에 무사히 5단계까지 올 수 있었습니다. 이번 피드백도 생각할 것들이 정말 많았고 스스로 완성됐다고 생각했던 코드에 얼마나 많은 스멜이 숨겨져있었는지 깨달았어요. 😭  더불어 나는 코틀린을 정말 모르는구나를 느꼈습니다 ㅎㅎ

도메인 로직과 UI 로직의 명확한 분리를 목표로 진행했습니다. ♻️ 

개선점이나 추가 의견이 있다면 적극 반영하겠습니다.
이번 PR도 잘 부탁드립니다! 🤩 

<img width="376" alt="image" src="https://github.com/user-attachments/assets/d25ebb40-4f7c-472c-bbed-305a110f589b">

Commit:
refactor: CarName을 Car의 내부 value class로 이동

- 자동차 이름 검증 로직을 Car 내부에서 처리

Commit:
refactor: Cars 코딩 컨벤션에 맞춰 내부 요소들을 정렬하라

Commit:
refactor: findWinners 메소드는 가독성을 증가시켜라

- 추출된 함수들을 함수체이닝을 통해 코드 통합

Commit:
refactor: Cars 생성시 부생성자를 사용하라

- 간단한 생성 로직을 부생성자로 처리하여 코드 간결화.
- 필요 시 다양한 생성 방식을 추가할 수 있도록 확장성 고려.

Commit:
refactor: 객체 이름과 메소드 명이 중복되지 않게 하라

Commit:
add: mockk 의존성을 추가하라

Commit:
refactor: 결과 출력 책임을 RaceGame에게 위임하라

Commit:
refactor: 우승자를 확인하는 책임을 RaceGame에 위임하라

Commit:
refactor: 우승자의 출력 로직을 뷰 계층에서 처리하라

Commit:
refactor: 자동차 생성시 위치를 지정하라

- 자동차 생성시 위치를 지정
- 테스트 유연성 증가

wisemuji commented on build.gradle.kts:
```diff
@@ -18,6 +18,7 @@ dependencies {
     testImplementation("org.junit.jupiter", "junit-jupiter", "5.10.2")
     testImplementation("org.assertj", "assertj-core", "3.25.3")
     testImplementation("io.kotest", "kotest-runner-junit5", "5.8.0")
+    testImplementation("io.mockk", "mockk", "1.13.13")
```
모킹을 사용하여 테스트를 만들어주셨군요! 하지만 구조만 잘 잡으면 모킹 없이도 충분히 테스트 가능한 구조를 만들 수 있습니다. 그리고 테스트 가능한 구조를 만들다 보면 자연스럽게 하나의 역할을 하는 객체 단위가 나눠지게 되어요. 이 교육과정 동안만큼은 모킹 없이 설계하는 연습을 해보면 어떨까요? :relaxed:

kyuwon53 replied:
좋습니다!

wisemuji commented on src/main/kotlin/app/Main.kt:
```diff
@@ -10,9 +10,9 @@ fun main() {
     ResultView.displayStartMessage()
-    raceGame.start()
-    ResultView.displayWinners(raceGame.winners())
+    val result = raceGame.run()
```
run이라는 네이밍은 Kotlin Scope Function 중 하나인 [run](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/run.html)과 헷갈릴 여지가 있을 것 같아요. 더 직관적인 네이밍으로 수정해보면 어떨까요?

wisemuji commented on src/main/kotlin/controller/RaceGame.kt:
```diff
@@ -4,15 +4,23 @@ import domain.CarNames
+    fun run(): Cars {
+        return (1..rounds).fold(initialCars) { currentCars, _ ->
+            val currentRace = Race(currentCars).start()
+            ResultView.displayRoundResults(currentCars)
```
마지막 단계인 만큼 좀 더 개선해볼 수 있는 여러 아이디어를 제시하고 싶어요 🙂

OutputView 호출을 최대한 줄여보면 어떨까요?

우리가 흔히 웹 프로그램에서 서버와 클라이언트 간에 소통하는걸 생각해보면
서버와 클라이언트가 서로 요청과 응답을 주고받는 과정은 모두 비용입니다. 이러한 비용을 줄이는 것이 웹 프로그래밍에서 중요하게 여겨지기도 합니다.

또 지금은 단순한 콘솔 프로그램이지만, 나중에 뷰가 안드로이드 앱으로 바뀐다던지, 웹페이지 등으로 바뀐다고 하면 이러한 호출이 모두 비용이 됩니다.

wisemuji commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -2,14 +2,23 @@ package domain
+
+    constructor(name: CarName, initialPosition: Int) : this(name, Position(initialPosition))
+
+    @JvmInline
+    value class CarName(val value: String) {
+        init {
+            require(value.isNotBlank()) { "자동차 이름은 공백일 수 없습니다." }
+            require(value.length <= 5) { "자동차 이름은 5자를 초과할 수 없습니다." }
+        }
+    }
```
class 안에 class를 넣은 형태로 작성해주신 특별한 이유가 있을까요?

wisemuji commented on src/main/kotlin/domain/Position.kt:
```diff
@@ -1,8 +1,6 @@
-class Position {
-    private var distance: Int = 0
-
+class Position(private var distance: Int = 0) {
```
distance() 함수는 position 값을 조회할 수 있는 일종의 getter인 것 같은데요,
"getter를 사용하는 대신 객체에 메시지를 보내자"라는 주제로 작성된 아래 글이 도움이 되면 좋겠습니다.

- [tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter)

제 의견은 항상 정답이 아닙니다. 규원님은 어떻게 생각하시나요?

wisemuji commented on src/main/kotlin/domain/RaceJudge.kt:
```diff
@@ -2,26 +2,17 @@ package domain
 
 class RaceJudge {
     fun findWinners(cars: Cars): Winners {
```
책 "객체지향의 사실과 오해"에서는 메세지에 대해 다음과 같이 설명합니다.

> 하나의 객체는 메시지를 전송함으로써 다른 객체에 접근한다(메시지-전송 메커니즘: 객체가 다른 객체에 접근할 수 있는 유일한 방법)
객체가 메시지를 수신할 수 있다는 것은 객체가 메시지에 해당하는 책임을 수행할 수 있다는 것이다.
외부의 객체는 메시지에 관해서만 볼 수 있고 객체 내부는 볼 수 없기 때문에 자연스럽게 객체의 외부와 내부가 분리된다.
메시지는 수신자,메시지 이름,인자 의 조합으로 이뤄진다.

`RaceJudge`라는 객체가 우승자를 판단하여 `Winners`를 대신 생성해주고 있는데요, 
`RaceJudge`, `Winners`과 같은 객체는 “자신의 상태를 스스로 관리하는 자율적인 존재“인 능동적인 객체라고 보기는 어렵다고 생각합니다. 무엇을 하는지에 강하게 묶여 있어 추가 책임을 가질 수도, 스스로 판단하여 메세지를 주고받을 수 있는 자유도 없습니다.

우승자를 판별하는 로직은 Winners 객체 내부가 스스로 판단하게 구현하고, 외부에서는 단순히 “우승자를 알려줘” 라는 메시지만 던지도록 개선해보면 어떨까요~?

다음 글이 도움이 되면 좋겠습니다.
- [객체지향의 사실과 오해 [4장. 역할, 책임, 협력]](https://github.com/Coveong/reading-books-for-programmers/blob/main/%EA%B0%9D%EC%B2%B4%EC%A7%80%ED%96%A5%EC%9D%98%20%EC%82%AC%EC%8B%A4%EA%B3%BC%20%EC%98%A4%ED%95%B4/04%EC%9E%A5.%20%EC%97%AD%ED%95%A0%2C%20%EC%B1%85%EC%9E%84%2C%20%ED%98%91%EB%A0%A5/wisemuji.md)
- [객체지향의 사실과 오해 [5장. 책임과 메시지]](https://github.com/Coveong/reading-books-for-programmers/blob/main/%EA%B0%9D%EC%B2%B4%EC%A7%80%ED%96%A5%EC%9D%98%20%EC%82%AC%EC%8B%A4%EA%B3%BC%20%EC%98%A4%ED%95%B4/05%EC%9E%A5.%20%EC%B1%85%EC%9E%84%EA%B3%BC%20%EB%A9%94%EC%8B%9C%EC%A7%80/wisemuji.md)

wisemuji reviewed (changes requested):
안녕하세요 규원님!

5단계 구현을 잘 해주셨습니다 💯 이전 단계 피드백 반영하시면서 코드 구조가 많이 개선된 것이 보입니다!
마지막 단계인 만큼 도전적인 코멘트를 몇 가지 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 알려주세요 ㅎㅎ

kyuwon53 committed:
refactor: RaceGame 시작 메소드 명을 수정하라

Commit:
remove: mockk 의존성을 제거하라

Commit:
refactor: 이동 판단 여부를 position으로 위임하라

- 이동 판단 여부 position으로 위임
- 이동 후 현재 상태 리턴

Commit:
refactor: 이동조건의 디폴트 값을 설정하라

Commit:
refactor: 경주 후 자동차 상태를 반환하라

Commit:
feat: 경기 별 기록을 저장하라

Commit:
feat: 경기 별 기록을 출력하라

Commit:
remove: Race를 제거하라

- RaceGame과 책임 중복

Commit:
refactor: 자동차들을 경주시키고 기록하라

- race 메시지를 자동차에게 전달
- 라운드 결과 기록
- 기록을 출력 view에 전달하여 한번에 출력

Commit:
refactor: Winners 객체가 스스로 우승자를 판별하라

- 우승자를 판별하는 책임을 Winners로 이동

kyuwon53 commented:
좀 더 고민하고 도전하고 있는 부분이 남아있지만, 우선 주신 피드백을 토대로 개선해본 코드 리뷰 요청드립니다.

wisemuji commented on src/main/kotlin/domain/RaceRecord.kt:
```diff
@@ -0,0 +1,10 @@
+package domain
+
+class RaceRecord(raceRecord: List<Cars> = emptyList()) {
+    private val _raceRecord = raceRecord.toMutableList()
+    val raceRecord: List<Cars> get() = _raceRecord
```
Immutable 타입은 외부에 공개하고, 내부에만 Mutable 타입을 공개하도록 잘 구현해주셨네요!

이미 아실 것 같지만 이러한 형태를 [backing property](https://kotlinlang.org/docs/properties.html#backing-properties)라고도 부릅니다.

wisemuji commented on src/main/kotlin/domain/Winners.kt:
```diff
@@ -2,4 +2,12 @@ package domain
+            return Winners(winningCars)
+        }
+    }
 }
```
피드백을 너무 잘 반영해주셨네요!

현재는 Winners를 생성하는 즉시 바로 우승자 이름이 만들어지게 되는데요, 부생성자나 팩토리 함수는 위에서 표현한 메세지의 일종으로 볼 수 있을지도 고민해보면 좋을 것 같아요. 현재 단계에서는 참고만 해주셔도 됩니다 🙂

wisemuji commented on src/main/kotlin/view/ResultView.kt:
```diff
@@ -1,14 +1,21 @@
-    fun displayRoundResults(cars: Cars) {
+    fun displayRoundResults(raceRecord: RaceRecord) {
+        raceRecord.raceRecord
+            .forEach(::displayRoundResult)
```
함수 reference를 활용해주셨네요! lambda로 작성하는 것과 어떤 차이가 있는지 참고할 수 있는 글을 소개해드릴게요 :slightly_smiling_face:
https://proandroiddev.com/kotlin-lambda-vs-method-reference-fdbd175f6845

참고로 저는 가능한 곳에서는 꼭 함수 reference를 활용하는 것을 지향합니다. lambda로 가져온 데이터를 별도로 조작하지 않고 그대로 함수 시그니처로 넘긴다는 것을 드러낼 수 있기 때문입니다.

wisemuji commented on src/test/kotlin/domain/PositionTest.kt:
```diff
@@ -1,25 +1,39 @@
+            Then("거리는 유지된다") {
+                position.distance() shouldBe 1
+            }
+        }
```
실행 순서가 보장되지 않는 환경에서는 Kotest의 [LifeCycle Hooks](https://kotest.io/docs/framework/lifecycle-hooks.html)를 활용해볼 수 있어요.
지금 상황에서는 beforeEach를 활용해볼 수 있을 것 같네요 🙂

wisemuji commented on src/test/kotlin/domain/WinnersTest.kt:
```diff
@@ -0,0 +1,35 @@
+import io.kotest.core.spec.style.BehaviorSpec
+import io.kotest.matchers.shouldBe
+
+class WinnersTest : BehaviorSpec({
```
적절한 시나리오를 잘 도출해주셨네요 👍 
테스트 시나리오 작성에 많이 익숙해지신 것 같아요! 앞으로의 미션도 기대됩니다

wisemuji reviewed (changes requested):
안녕하세요 규원님! 리뷰 반영을 정말 잘 해주셨습니다 💯 
객체 간의 역할/책임도 잘 정리되었고 코드 구조가 많이 개선되었어요.

조금 더 추가로 고민해보시면 좋을 내용을 코멘트로 달았는데요,
미션을 종료하고 싶으시다면 다시 리뷰 요청 해주세요. 
다음 리뷰 요청에서는 PR을 머지할 수 있을 것 같아요 🙂

kyuwon53 committed:
refactor: Winners 클래스 리팩토링 및 determine 메서드 개선

- Winners가 Cars를 내부 상태로 보유하도록 변경
- determine 메서드가 파라미터 없이 내부 상태를 기반으로 동작
- lazy를 사용해 우승자 계산을 지연 실행하도록 구현
- 불변성을 유지하며 객체지향 설계를 강화

kyuwon53 committed:
remove: 사용하지 않는 임포트를 제거하라

kyuwon53 committed:
refactor: 테스트 후 테스트 스텁을 초기화하라

kyuwon53 commented:
주신 피드백 반영했습니다! 
회사가 계속 바빠서 그런지 ;;; 시간이 안나네요 😢 
position에서 게터를 제거하는 시도중입니다...ㅎㅎ
혹시 계속해서 리뷰 요청을 드려도되나요?

kyuwon53 committed:
refactor: Position과 CarName을 Car에게 위임하라

- Position CarName의 책임은 car에게 있다

kyuwon53 committed:
refactor: 우승자 판별할 때 자동차에 메시지를 보내라

- Winners 클래스에서 우승자 판별 로직을 간소화
  - maxOf와 filter를 활용하여 자동차 간 비교를 명확히 표현
  - 자동차의 위치 비교를 Car 클래스의 메시지 호출로 캡슐화
- Car 클래스에 isAtDistance 메서드 추가
  - 특정 거리와의 비교 로직을 캡슐화하여 메시지 기반 설계를 강화
- 객체지향 원칙에 따라 메시지를 통해 협력하도록 구조 개선

kyuwon53 committed:
refactor: Car의 currentDistance 제거

wisemuji commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,22 +1,29 @@
-    fun currentDistance(): Int = position.distance()
-
-    constructor(name: CarName, initialPosition: Int) : this(name, Position(initialPosition))
+    var position: Int = position
+        private set
```
생성자로 초기 값 주입이 가능해졌지만, 외부에서 수정할 수는 없도록 잘 막아주셨네요 👍

wisemuji commented on src/main/kotlin/domain/Car.kt:
```diff
@@ -1,22 +1,29 @@
+
+    fun hasSamePositionAs(other: Car): Boolean {
+        return this.position == other.position
     }
```
Comparable interface를 구현하는 것도 고려해볼 수 있을 것 같아요! 꼭 반영하진 않으셔도 됩니다 🙂 

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-comparable/

wisemuji commented on src/main/kotlin/domain/CarNames.kt:
```diff
@@ -1,12 +1,10 @@
-        require(values.isNotEmpty()) { "자동차 이름은 하나 이상이어야 합니다." }
-        require(values.distinct().size == values.size) { "자동차 이름은 중복될 수 없습니다." }
+        require(names.isNotEmpty()) { "자동차 이름은 하나 이상이어야 합니다." }
+        require(names.distinct().size == names.size) { "자동차 이름은 중복될 수 없습니다." }
```
require와 비슷한 [check](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/check.html)라는 API도 있는데요, 혹시 require와 check는 각각 어떤 상황에서 사용하는 것이 좋다고 생각하시나요?

wisemuji replied:
중복을 허용하지 않는 [Set](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-set/) 타입을 활용해보는 것도 고려해볼 수 있을 것 같아요 🙂  반영 여부는 규원님이 판단해주세요!

wisemuji reviewed (changes requested):
안녕하세요 규원님! 추가 피드백도 잘 반영해 주셨습니다 💯 

> 혹시 계속해서 리뷰 요청을 드려도되나요?

아직 미션을 종료하고 싶지 않으신 것 같아서 한 번 더 RC를 드렸는데요!
혹시 추가로 도전해보고 싶으신 내용이 있으실까요~?

kyuwon53 committed:
refactor: 자동차 이름 중복 제거 방식을 수정하라

- set을 사용하여 중복 제거

kyuwon53 committed:
refactor: 제거된 게터 함수 테스트에 적용

kyuwon53 committed:
refactor: Comparable을 사용하여 위치를 비교하라

kyuwon53 commented:
> position에서 게터를 제거하는 시도중입니다...ㅎㅎ

1주차 안에 저 포지션 부분을 못 끝낼거 같아서 리뷰를 계속 할 수 있는지 여쭤봤습니다 😅 
너무 잘 리뷰해주셔서 많이 배웠습니다. 감사합니다 🚀 👍

wisemuji reviewed (approved):
안녕하세요 규원님! 마지막까지 리뷰 반영 하시느라 고생이 많으셨습니다.

규원님과 의견을 주고받으며 저 또한 좋은 인사이트를 얻어가는 시간이였습니다!
개발자로 성장하는 과정에 도움이 되었길 바라며 이번 미션은 마무리하겠습니다.
나머지 미션도 꼭 완주하시길 응원하겠습니다 💪
